pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java

```diff
@@ -140,12 +141,23 @@ public void initChannel(SocketChannel ch) throws Exception {
 
             future.channel().closeFuture().addListener(v -> {
                 // Remove connection from pool when it gets closed
+                if (log.isDebugEnabled()) {
+                    log.debug("Removing closed connection from pool: {}", v);
+                }
                 cleanupConnection(address, connectionKey, cnxFuture);
             });
 
             // We are connected to broker, but need to wait until the connect/connected handshake is
             // complete
             final ClientCnx cnx = (ClientCnx) future.channel().pipeline().get("handler");
+            if (!future.channel().isActive() || cnx == null) {
```

Comment:
The problem is that the connection was actually successfully open, thus the future itself is completed, but then immediately closed by the broker (that's what the test is reproducing).

So the future still stays set to success (future cannot change their status once they flip), but the isActive() now returns false, and the "handler" is also removed from the channel pipeline, causing the NPE when accessing cnx.